File: services/fiscal-engine/app/product_reconciliation.py
from __future__ import annotations

from collections import defaultdict
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP
import re
import unicodedata


ZERO = Decimal("0")
CENT = Decimal("0.01")


def decimal_value(value, default: Decimal = ZERO) -> Decimal:
    try:
        return Decimal(str(value)).quantize(Decimal("0.000001"))
    except (InvalidOperation, TypeError, ValueError):
        return default


def money(value: Decimal) -> Decimal:
    return value.quantize(CENT, rounding=ROUND_HALF_UP)


def _normalized(value: str | None) -> str:
    raw = unicodedata.normalize("NFKD", str(value or ""))
    ascii_value = "".join(character for character in raw if not unicodedata.combining(character))
    return re.sub(r"[^A-Z0-9]+", " ", ascii_value.upper()).strip()


def _valid_gtin(value: str | None) -> str | None:
    digits = re.sub(r"\D", "", str(value or ""))
    return digits if len(digits) in {8, 12, 13, 14} else None


def item_identity(item: dict) -> dict | None:
    details = item.get("details") or {}
    configured = details.get("reconciliation_identity") or {}
    if configured.get("key") and configured.get("label"):
        return configured

    if item.get("chassis"):
        chassis = _normalized(item["chassis"]).replace(" ", "")
        return {
            "key": f"unique:chassis:{chassis}",
            "label": f"Chassi {item['chassis']}",
            "type": "chassis",
            "confidence": "exact",
            "basis": "det/infAdProd",
        }

    identifiers = details.get("identifiers") or []
    priorities = {"chassis": 0, "imei": 1, "serial": 2, "aggregation_code": 3}
    unique = sorted(
        (identifier for identifier in identifiers if identifier.get("type") in priorities and identifier.get("value")),
        key=lambda identifier: priorities[identifier["type"]],
    )
    if unique:
        identifier = unique[0]
        kind = identifier["type"]
        value = _normalized(identifier["value"]).replace(" ", "")
        labels = {"chassis": "Chassi", "imei": "IMEI", "serial": "Série", "aggregation_code": "Agregação"}
        return {
            "key": f"unique:{kind}:{value}",
            "label": f"{labels[kind]} {identifier['value']}",
            "type": kind,
            "confidence": "exact",
            "basis": identifier.get("source"),
        }

    traceability = details.get("traceability") or []
    lot = next((entry for entry in traceability if entry.get("lot")), None)
    if lot:
        qualifier = _valid_gtin(details.get("ean_taxable") or details.get("ean")) or item.get("ncm") or "SEM-CODIGO"
        return {
            "key": f"lot:{qualifier}:{_normalized(lot['lot']).replace(' ', '')}",
            "label": f"Lote {lot['lot']} · {item.get('description') or qualifier}",
            "type": "lot",
            "confidence": "high",
            "basis": "prod/rastro/nLote",
        }

    gtin = _valid_gtin(details.get("ean_taxable") or details.get("ean"))
    if gtin:
        return {
            "key": f"gtin:{gtin}",
            "label": f"GTIN {gtin} · {item.get('description') or ''}".rstrip(" ·"),
            "type": "gtin",
            "confidence": "high",
            "basis": "prod/cEAN|cEANTrib",
        }

    description = _normalized(item.get("description"))
    ncm = re.sub(r"\D", "", str(item.get("ncm") or ""))
    unit = _normalized(details.get("unit") or details.get("taxable_unit"))
    if ncm and description:
        return {
            "key": f"indicative:{ncm}:{description}:{unit}",
            "label": item.get("description") or ncm,
            "type": "ncm_description",
            "confidence": "indicative",
            "basis": "prod/NCM+xProd+uCom",
        }

    return None


def build_product_reconciliation(documents: list[dict]) -> list[dict]:
    grouped: dict[str, list[tuple[dict, dict, dict]]] = defaultdict(list)
    for document in documents:
        if document.get("status") == "cancelled":
            continue
        for item in document.get("items", []):
            identity = item_identity(item)
            if identity:
                grouped[identity["key"]].append((document, item, identity))

    rows = []
    for key, entries in grouped.items():
        identity = entries[0][2]
        inputs = [entry for entry in entries if entry[0].get("direction") == "entrada"]
        outputs = [entry for entry in entries if entry[0].get("direction") == "saida"]
        input_quantity = sum((_quantity(item, identity) for _, item, _ in inputs), ZERO)
        output_quantity = sum((_quantity(item, identity) for _, item, _ in outputs), ZERO)
        input_value = money(sum((decimal_value(item.get("product_value")) for _, item, _ in inputs), ZERO))
        output_value = money(sum((decimal_value(item.get("product_value")) for _, item, _ in outputs), ZERO))
        quantity_complete = all(_has_quantity(item, identity) for _, item, _ in inputs + outputs)
        unit_cost = money(input_value / input_quantity) if input_quantity > ZERO else None
        enough_stock = input_quantity >= output_quantity > ZERO
        estimated_cost = money(unit_cost * output_quantity) if unit_cost is not None and quantity_complete and enough_stock else None
        margin = money(output_value - estimated_cost) if estimated_cost is not None else None

        if not outputs:
            status = "in_stock"
        elif not inputs:
            status = "missing_input"
        elif not quantity_complete:
            status = "insufficient_quantity_data"
        elif not enough_stock:
            status = "insufficient_input_quantity"
        elif identity["confidence"] == "indicative":
            status = "review_identity"
        elif margin is not None and margin < ZERO:
            status = "negative_margin"
        elif margin == ZERO:
            status = "zero_margin"
        else:
            status = "reconciled" if identity["confidence"] == "exact" else "reconciled_estimate"

        latest_input = max(inputs, key=lambda entry: entry[0].get("issued_at") or "", default=None)
        latest_output = max(outputs, key=lambda entry: entry[0].get("issued_at") or "", default=None)
        rows.append({
            "key": key,
            "identifier": identity["label"],
            "identity_type": identity["type"],
            "confidence": identity["confidence"],
            "basis": identity.get("basis"),
            "description": next((item.get("description") for _, item, _ in entries if item.get("description")), None),
            "ncm": next((item.get("ncm") for _, item, _ in entries if item.get("ncm")), None),
            "unit": next(((item.get("details") or {}).get("unit") for _, item, _ in entries if (item.get("details") or {}).get("unit")), None),
            "input_quantity": str(input_quantity.normalize()) if input_quantity else "0",
            "output_quantity": str(output_quantity.normalize()) if output_quantity else "0",
            "input_value": str(input_value),
            "output_value": str(output_value),
            "estimated_unit_cost": str(unit_cost) if unit_cost is not None else None,
            "estimated_cost": str(estimated_cost) if estimated_cost is not None else None,
            "margin": str(margin) if margin is not None else None,
            "status": status,
            "input_document": latest_input[0].get("number") if latest_input else None,
            "output_document": latest_output[0].get("number") if latest_output else None,
            "output_document_ref": latest_output[0].get("document_ref") if latest_output else None,
            "output_item_number": latest_output[1].get("item_number") if latest_output else None,
            "used_movable_good": any(bool(item.get("used_movable_good")) for _, item, _ in entries),
            "chassis": next((item.get("chassis") for _, item, _ in entries if item.get("chassis")), None),
        })

    confidence_order = {"exact": 0, "high": 1, "indicative": 2}
    return sorted(rows, key=lambda row: (confidence_order.get(row["confidence"], 9), row["identifier"]))


def _has_quantity(item: dict, identity: dict) -> bool:
    if identity.get("confidence") == "exact":
        return True
    details = item.get("details") or {}
    return decimal_value(details.get("quantity")) > ZERO


def _quantity(item: dict, identity: dict) -> Decimal:
    if identity.get("confidence") == "exact":
        return Decimal("1")
    details = item.get("details") or {}
    return decimal_value(details.get("quantity"))

File: apps/api/app/Services/ProductReconciliation.php
<?php

namespace App\Services;

use Illuminate\Support\Str;

class ProductReconciliation
{
    public function build(iterable $items): array
    {
        $groups = [];
        foreach ($items as $item) {
            $identity = $this->identity($item);
            if (! $identity) {
                continue;
            }
            $groups[$identity['key']][] = ['item' => $item, 'identity' => $identity];
        }

        $rows = [];
        foreach ($groups as $key => $entries) {
            $identity = $entries[0]['identity'];
            $inputs = array_values(array_filter($entries, fn (array $entry): bool => $entry['item']->document?->direction === 'entrada'));
            $outputs = array_values(array_filter($entries, fn (array $entry): bool => $entry['item']->document?->direction === 'saida'));
            $inputQuantity = $this->sumQuantity($inputs, $identity);
            $outputQuantity = $this->sumQuantity($outputs, $identity);
            $inputValue = $this->sumMoney($inputs);
            $outputValue = $this->sumMoney($outputs);
            $quantityComplete = collect([...$inputs, ...$outputs])->every(
                fn (array $entry): bool => $this->hasQuantity($entry['item'], $identity),
            );
            $unitCost = bccomp($inputQuantity, '0', 6) > 0 ? bcdiv($inputValue, $inputQuantity, 6) : null;
            $enoughStock = bccomp($outputQuantity, '0', 6) > 0 && bccomp($inputQuantity, $outputQuantity, 6) >= 0;
            $estimatedCost = $unitCost !== null && $quantityComplete && $enoughStock
                ? $this->roundMoney(bcmul($unitCost, $outputQuantity, 6))
                : null;
            $margin = $estimatedCost !== null ? bcsub($outputValue, $estimatedCost, 2) : null;

            $status = match (true) {
                $outputs === [] => 'in_stock',
                $inputs === [] => 'missing_input',
                ! $quantityComplete => 'insufficient_quantity_data',
                ! $enoughStock => 'insufficient_input_quantity',
                $identity['confidence'] === 'indicative' => 'review_identity',
                $margin !== null && bccomp($margin, '0', 2) < 0 => 'negative_margin',
                $margin !== null && bccomp($margin, '0', 2) === 0 => 'zero_margin',
                $identity['confidence'] === 'exact' => 'reconciled',
                default => 'reconciled_estimate',
            };
            $latestInput = $this->latest($inputs);
            $latestOutput = $this->latest($outputs);
            $firstItem = $entries[0]['item'];

            $rows[] = [
                'key' => $key,
                'identifier' => $identity['label'],
                'identity_type' => $identity['type'],
                'confidence' => $identity['confidence'],
                'basis' => $identity['basis'] ?? null,
                'description' => $firstItem->description ?? null,
                'ncm' => $firstItem->ncm ?? null,
                'unit' => data_get($firstItem->details, 'unit'),
                'input_quantity' => $this->decimalText($inputQuantity),
                'output_quantity' => $this->decimalText($outputQuantity),
                'input_value' => $inputValue,
                'output_value' => $outputValue,
                'estimated_unit_cost' => $unitCost,
                'estimated_cost' => $estimatedCost,
                'margin' => $margin,
                'status' => $status,
                'input_document' => $latestInput['item']->document?->number ?? null,
                'output_document' => $latestOutput['item']->document?->number ?? null,
                'chassis' => $firstItem->chassis ?? null,
            ];
        }

        $confidence = ['exact' => 0, 'high' => 1, 'indicative' => 2];
        usort($rows, fn (array $left, array $right): int =>
            [$confidence[$left['confidence']] ?? 9, $left['identifier']]
            <=> [$confidence[$right['confidence']] ?? 9, $right['identifier']]
        );

        return $rows;
    }

    private function identity(object $item): ?array
    {
        $details = is_array($item->details) ? $item->details : [];
        $configured = $details['reconciliation_identity'] ?? null;
        if (is_array($configured) && ! empty($configured['key']) && ! empty($configured['label'])) {
            return $configured;
        }
        if ($item->chassis ?? null) {
            return ['key' => 'unique:chassis:'.$this->compact($item->chassis), 'label' => 'Chassi '.$item->chassis, 'type' => 'chassis', 'confidence' => 'exact', 'basis' => 'det/infAdProd'];
        }

        $priority = ['chassis' => 0, 'imei' => 1, 'serial' => 2, 'aggregation_code' => 3];
        $identifiers = array_values(array_filter($details['identifiers'] ?? [], fn ($identifier): bool =>
            is_array($identifier) && isset($priority[$identifier['type'] ?? '']) && ! empty($identifier['value'])
        ));
        usort($identifiers, fn (array $left, array $right): int => $priority[$left['type']] <=> $priority[$right['type']]);
        if ($identifiers !== []) {
            $identifier = $identifiers[0];
            $labels = ['chassis' => 'Chassi', 'imei' => 'IMEI', 'serial' => 'Série', 'aggregation_code' => 'Agregação'];
            return ['key' => "unique:{$identifier['type']}:{$this->compact($identifier['value'])}", 'label' => "{$labels[$identifier['type']]} {$identifier['value']}", 'type' => $identifier['type'], 'confidence' => 'exact', 'basis' => $identifier['source'] ?? null];
        }

        $lot = collect($details['traceability'] ?? [])->first(fn ($entry): bool => is_array($entry) && ! empty($entry['lot']));
        if ($lot) {
            $qualifier = $this->validGtin($details['ean_taxable'] ?? $details['ean'] ?? null) ?: (($item->ncm ?? null) ?: 'SEM-CODIGO');
            return ['key' => "lot:{$qualifier}:{$this->compact($lot['lot'])}", 'label' => "Lote {$lot['lot']} · ".(($item->description ?? null) ?: $qualifier), 'type' => 'lot', 'confidence' => 'high', 'basis' => 'prod/rastro/nLote'];
        }

        $gtin = $this->validGtin($details['ean_taxable'] ?? $details['ean'] ?? null);
        if ($gtin) {
            return ['key' => "gtin:{$gtin}", 'label' => 'GTIN '.$gtin.(($item->description ?? null) ? " · {$item->description}" : ''), 'type' => 'gtin', 'confidence' => 'high', 'basis' => 'prod/cEAN|cEANTrib'];
        }

        $description = $this->normalized($item->description ?? null);
        $ncm = preg_replace('/\D/', '', (string) ($item->ncm ?? ''));
        $unit = $this->normalized($details['unit'] ?? $details['taxable_unit'] ?? null);
        if ($ncm !== '' && $description !== '') {
            return ['key' => "indicative:{$ncm}:{$description}:{$unit}", 'label' => ($item->description ?? null) ?: $ncm, 'type' => 'ncm_description', 'confidence' => 'indicative', 'basis' => 'prod/NCM+xProd+uCom'];
        }

        return null;
    }

    private function sumQuantity(array $entries, array $identity): string
    {
        return array_reduce($entries, fn (string $carry, array $entry): string =>
            bcadd($carry, $this->quantity($entry['item'], $identity), 6), '0.000000');
    }

    private function sumMoney(array $entries): string
    {
        return array_reduce($entries, fn (string $carry, array $entry): string =>
            bcadd($carry, (string) ($entry['item']->product_value ?? '0'), 2), '0.00');
    }

    private function quantity(object $item, array $identity): string
    {
        return $identity['confidence'] === 'exact' ? '1.000000' : (string) (data_get($item->details, 'quantity') ?: '0');
    }

    private function hasQuantity(object $item, array $identity): bool
    {
        return $identity['confidence'] === 'exact' || bccomp($this->quantity($item, $identity), '0', 6) > 0;
    }

    private function latest(array $entries): ?array
    {
        if ($entries === []) {
            return null;
        }
        usort($entries, fn (array $left, array $right): int =>
            ($left['item']->document?->issued_at?->getTimestamp() ?? 0) <=> ($right['item']->document?->issued_at?->getTimestamp() ?? 0)
        );
        return end($entries) ?: null;
    }

    private function roundMoney(string $value): string
    {
        return bcadd($value, str_starts_with($value, '-') ? '-0.005' : '0.005', 2);
    }

    private function decimalText(string $value): string
    {
        $text = str_contains($value, '.') ? rtrim(rtrim($value, '0'), '.') : $value;
        return $text === '' || $text === '-0' ? '0' : $text;
    }

    private function validGtin(?string $value): ?string
    {
        $digits = preg_replace('/\D/', '', (string) $value);
        return in_array(strlen($digits), [8, 12, 13, 14], true) ? $digits : null;
    }

    private function compact(string $value): string
    {
        return preg_replace('/[^A-Z0-9]+/', '', Str::upper(Str::ascii($value)));
    }

    private function normalized(?string $value): string
    {
        return trim(preg_replace('/[^A-Z0-9]+/', ' ', Str::upper(Str::ascii((string) $value))));
    }
}
